doc: ipc: Add detailed protocol specification for ICBMsg backend (#77229)
zephyrproject-rtos/zephyr: doc/services/ipc/ipc_service/backends/ipc_service_icbmsg.rst

=======

* :ref:`ipc_multi_endpoint_sample`

> Detailed Protocol Specification

Author: I disagree. The "Detailed Protocol Specification" is for users that need to know low level details about the protocol. The configuration is for all user. They do not need to know how the `rx-region` and `tx-region` is constructed. They just need to assign a memory for them using configuration.

===============================

Shared Memory Organization
--------------------------

The ICBMsg uses two shared memory regions, ``rx-region`` for message receiving, and ``tx-region`` for message transmission.

> The regions do not have to be next to each other nor placed in any specification order nor have the same size.

Reviewer: ```suggestion
The regions do not have to be next to each other nor placed in any specific order nor have the same size.
```

Reviewer 2: I would add a remark that these areas must be aligned to cache boundary and do not overlap.
...
I went further and see there is a remark regarding cache so  maybe this is not needed.

> Those regions are interchanged on both sides.

Author: No, not "interchangeable". I was trying to say that, for example:

If, on APP core, the `rx-region` is at address 1000 and `tx-region` is at address 5000,
then, on RADIO core, the `rx-region` must be at address 5000, `tx-region` must be at address 1000.

Each shared memory region is divided into two parts:

* **ICMsg area** - an area reserved by ICMsg instance and used for control messages.
* **Blocks area** - an area containing allocatable blocks carrying the content of the messages.
  This area is divided into even-sized blocks aligned to cache boundaries.

The location of each area is calculated to fulfill cache boundary requirements and allow optimal region usage.

> It is calculated using the following algorithm.

Reviewer: ```suggestion
It is calculated using the following algorithm:
```

Inputs:

* ``region_begin``, ``region_end`` - boundaries of the region,
* ``local_blocks`` - number of blocks in this region,
* ``remote_blocks`` - number of blocks in the opposite region,
* ``alignment`` - memory cache alignment.

The algorithm:

> #. Align region boundaries to cache:

Author: Keeping `#.` as discussed off-line.

When this control message is received the blocks containing the message must be released.

MSG_BOUND
^^^^^^^^^

.. list-table::
   :header-rows: 1

   * - byte 0
     - byte 1
     - byte 2
   * - MSG_BOUND
     - endpoint address
     - block number
   * - 0x02
     - 0x00 ÷ 0xFD
     - 0x00 ÷ N-1

The ``MSG_BOUND`` control message is similar to the ``MSG_DATA`` except the blocks carry bounding information.
See the next section for details on the bounding procedure.

MSG_RELEASE_BOUND
^^^^^^^^^^^^^^^^^

.. list-table::
   :header-rows: 1

   * - byte 0
     - byte 1
     - byte 2
   * - MSG_RELEASE_BOUND
     - endpoint address
     - block number
   * - 0x03
     - 0x00 ÷ 0xFD
     - 0x00 ÷ N-1

The ``MSG_RELEASE_BOUND`` control message is sent in response to ``MSG_BOUND``.
It is similar to the ``MSG_RELEASE_DATA`` except the ``endpoint address`` is required.
See the next section for details on the bounding procedure.

Initialization
--------------

The ICBMsg initialization calls ICMsg initialization.
When it is done, no further initialization is required.
Blocks area can be left uninitialized.

After ICBMsg initialization, we are ready for the endpoint bounding procedure.

> Endpoint Bounding

Reviewer 3: I am not sure, just checking and don't want to revolutionize the whole page but shouldn't it be _binding_ instead of _bounding_? Technically, the procedure is starting a connection between endpoints on different cores. Do you have an opinion @divipillai?

Reviewer 4: yeah, agree with you Michal. We should be using binding in this scenario, since we are talking about endpoints. Checked for similar example and in NCS, we use the term binding: https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/libraries/zigbee/shell.html#zdo_mgmtbind. 

_The correct term should be "binding message" when referring to messages exchanged during the process of establishing a connection between two endpoints in a network. The process itself should be referred to as "endpoint binding." The term "bounding" is not typically used in this context._ - From an another source.

Reviewer 4: Had discussed offline with @doki-nordic , and agreed to change to binding, where ever applicable. Rest docs looks good.

Initiator Bounding Procedure
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The initiator sends a bounding message.
It contains a single null-terminated string with an endpoint name.
As usual, it is preceded by a message header containing message size (including null-terminator).

Example of the bounding message for ``example`` endpoint name.

.. list-table::
   :header-rows: 1

   * - Header
     - Endpoint name
     - Null-terminator
   * - bytes 0-3
     - bytes 4-10
     - byte 11
   * - 0x00000008
     - ``example``
     - 0x00

The bounding message is sent using the ``MSG_BOUND`` control message and released with the ``MSG_RELEASE_BOUND`` control message.

The endpoint bounding procedure from the initiator's point of view is the following:

#. The initiator assigns an endpoint address to this endpoint.
#. The initiator sends a bounding message containing the endpoint name and address.
#. The initiator waits for any message from the follower using this endpoint address.
   Usually, it will be ``MSG_RELEASE_BOUND``, but ``MSG_DATA`` is also allowed.
#. The initiator is bounded and it can send data messages using this endpoint.

Follower Bounding Procedure
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> If the follower receives a bounding message before it starts the bounding procedure on that endpoint,
> it should store the message for later.

Reviewer 4: ```suggestion
If the follower receives a bounding message before it starts the bounding procedure on that endpoint, it should store the message for later.
```

There are no restrictions on the order in which the sides start the endpoint bounding.

Initiator Bounding Procedure
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The initiator sends a bounding message.
It contains a single null-terminated string with an endpoint name.
As usual, it is preceded by a message header containing the message size (including null-terminator).

Example of the bounding message for ``example`` endpoint name:

.. list-table::
   :header-rows: 1

   * - Header
     - Endpoint name
     - Null-terminator
   * - bytes 0-3
     - bytes 4-10
     - byte 11
   * - 0x00000008
     - ``example``
     - 0x00

The bounding message is sent using the ``MSG_BOUND`` control message and released with the ``MSG_RELEASE_BOUND`` control message.

The endpoint bounding procedure from the initiator's point of view is the following:

#. The initiator assigns an endpoint address to this endpoint.
#. The initiator sends a bounding message containing the endpoint name and address.
#. The initiator waits for any message from the follower using this endpoint address.
   Usually, it will be ``MSG_RELEASE_BOUND``, but ``MSG_DATA`` is also allowed.

> #. The initiator is bounded and it can send data messages using this endpoint.

Reviewer 4: ```suggestion
#. The initiator is bound to an endpoint, and it can send data messages using this endpoint.
```

Follower Bounding Procedure
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If the follower receives a bounding message before it starts the bounding procedure on that endpoint, it should store the message for later.
It should not send the ``MSG_RELEASE_BOUND`` yet.

The endpoint bounding procedure from the follower's point of view is the following:

#. The follower waits for a bounding message containing its endpoint name.
   The message may be a newly received message or a message stored before the bounding procedure started.
#. The follower stores the endpoint address assigned to this endpoint by the initiator.
#. The follower sends the ``MSG_RELEASE_BOUND`` control message.

> #. The follower is bounded and it can send data messages using this endpoint.

Reviewer 4: ```suggestion
#. The follower is bound to an endpoint, and it can send data messages using this endpoint.
```
